source/common/event/file_event_impl.h

```diff
@@ -41,5 +45,136 @@ class FileEventImpl : public FileEvent, ImplBase {
   // polling and activating new fd events.
   const bool activate_fd_events_next_event_loop_;
 };
+
+// Forward declare for friend class.
+class UserSpaceFileEventFactory;
+
+// The interface of populating event watcher and obtaining the active events. The events are the
+// combination of FileReadyType values. The event listener is populated by user event registration
+// and io events passively. Also the owner of this listener query the activated events by calling
+// triggeredEvents and getAndClearEphemeralEvents.
+class EventListener {
+public:
+  virtual ~EventListener() = default;
+
+  // Get the events which are enabled and triggered.
+  virtual uint32_t triggeredEvents() PURE;
+  // Get the events which are ephemerally activated. Upon returning the ephemeral events are
+  // cleared.
+  virtual uint32_t getAndClearEphemeralEvents() PURE;
+
+  /**
+   * FileEvent::setEnabled is invoked.
+   * @param enabled_events supplied the event of setEnabled.
+   */
+  virtual void onEventEnabled(uint32_t enabled_events) PURE;
+
+  /**
+   * FileEvent::activate is invoked.
+   * @param enabled_events supplied the event of activate().
+   */
+  virtual void onEventActivated(uint32_t activated_events) PURE;
+};
+
+// Return the enabled events except EV_CLOSED. This implementation is generally good since only
+// epoll supports EV_CLOSED but the entire envoy code base supports another poller. The event owner
+// must assume EV_CLOSED is never activated. Also event owner must tolerate that OS could notify
+// events which are not actually triggered.
+// TODO(lambdai): Add support of delivering EV_CLOSED.
+class DefaultEventListener : public EventListener {
+public:
+  ~DefaultEventListener() override = default;
+
+  // Return both read and write.
+  uint32_t triggeredEvents() override { return pending_events_ & (~Event::FileReadyType::Closed); }
+
+  void onEventEnabled(uint32_t enabled_events) override { pending_events_ = enabled_events; }
+
+  void onEventActivated(uint32_t activated_events) override {
+    ephermal_events_ |= activated_events;
+  }
+
+  uint32_t getAndClearEphemeralEvents() override { return std::exchange(ephermal_events_, 0); }
+
+private:
+  // The persisted interested events and ready events.
+  uint32_t pending_events_{};
```

Comment:
I find libevent terminology to be confusing. I would prefer `watched_events_` and `activated_events_` or `triggered_events_`. Up to you though. However the `triggeredEvents()` method is definitely a misnomer as it returns watched events without the `Closed` event.

Reply:
renamed to `enabled_events_`. This matches Envoy's dictionary. 

Closed event is another story. I think Envoy should survive without a Closed. But I will fixed it along with the efforts to deliver accurater events.